### PR TITLE
Feature/adding method sequence substract

### PR DESCRIPTION
**Pull request by Bibendus83**
Adding the new method Sequence::substract.
This function can be really useful in calendar operations when you want to substract a subset of intervals from another bigger interval.
Here is an example:

immagine

This feature should close issue #78

### CODE REVIEW

**Comment by nyamsprod** [comment]
I have a lot or questions 😂 i'll review your PR later this night when I have time.

**Comment by Bibendus83** [comment]
I'm not proud of the readability of this code but it was hell to create because the Period::substract method returns a Sequence instance.
Now that the function works and my tests are in place it should be easier to simplify it.

**Comment by nyamsprod** [comment]
My main complaint is not the code per se but the method signature... Why not remove the submitted input like Sequence::gap for instance

**Comment by nyamsprod** [comment]
Also I have the feeling that you can use Sequence::intersections to simplify the code 🤔🤔

**Comment by Bibendus83** [comment]
I don't get what do you mean with the removal of submitted input.

The intersection can't be used in substraction because it could remove useful data when, for instance, the substracted Sequence is empty or not overlapping.
For instance:
(10 elements) - (0 elements) = (10 elements)
(10 elements) ∩ (0 elements) = (0 elements)

**Comment by nyamsprod** [comment]
You take a sequence...
You calculate the sequence intersections
You "substract" the returned intersections sequence from the original one 
You get your substraction... I think

**Comment by Bibendus83** [comment]
I think it should work.
I shoud say that with my implementation the algorithm could reach a max complexity of N x M where N and M are the number of Periods in each Sequence.
Adding the intersection could reduce the amount of Periods (M) reducing the complexity in some cases.

**Comment by Bibendus83** [comment]
Mmmmm I noticed that the actual `Sequence::intersect` function is not working as I expected.
First the interesection is done only inside the sequence while in our case we would need an intersection between 2 Sequences.
Second the intersection should work taking all the common ranges in all contained Period, not any intersection between at least 2 Periods.

Example:
immagine
Should return empty instead

It should instead work like this:
immagine

**Comment by nyamsprod** [comment]
Sorry I haven't have time to review your founding 😥 If you coule proceed I'll see what I can do in the meantime

**Comment by Bibendus83** [comment]
what should I do? :)

**Comment by nyamsprod** [comment]
Try to implement it as n'est as you can we will refactor/improve it after don't worry

**Comment by Bibendus83** [comment]
What do you mean by n'est?

**Comment by nyamsprod** [comment]
Sorry my Smartphone is set in French 😒😒 I meant as good as you can

**Comment by Bibendus83** [comment]
I could add some comments but atm I don't know how to improve it, consider I already spent 2 days on it for the first implementation.

**Comment by nyamsprod** [comment]
I could merge it but I'll work on it on april i'm afraid

**Comment by Bibendus83** [comment]
I'm not in hurry, don't worry.
I'm already using this branch on my project.

**Comment by nyamsprod** [comment]
Here's an improved implementation... feel free to update your PR with it... 
```php
    public function substract(Sequence $sequence): self
    {
        return $sequence->reduce([$this, 'substractOne'], $this);
    }

    private function substractOne(Sequence $sequence, Period $interval): self
    {
        $new = new Sequence();
        foreach ($sequence as $period) {
            $substract = $period->substract($interval);
            if ($substract->isEmpty()) {
                continue;
            }

            $new->push(...$substract);
        }

        return $new;
    }
```

IMHO this one is easier to understand. Tell me what you think.

**Comment by nyamsprod** [comment]
I've updated the documentation website and used your image to illustrate the method usage 😉. Could you review the documentation and the improved implementation. Depending on your feedback I may release v4.7 next week

**Comment by Bibendus83** [comment]
Sure not it's more readable.
I have some doubts about the use of boundaries()->overlaps(), because it's a weak control that would often return true increasing calculation time for low benefits.
Another thing I don't understand is the use of reduce(), doesn't it reduce the whole sequence to a single Period?

**Comment by nyamsprod** [comment]
the boundaries overlaps could be dropped indeed ... it's kind of a tradeoff between performance because the code is full of O(n) operations which are bad for performance.

Sequence::reduce returns what you want it to return 😄 . Depending on the given callable it can returns a scalar/Object/array/Period/Sequence .... that what cool about it 😉

**Comment by Bibendus83** [comment]
What I feel is wrong is the use of reduce() called on the $sequence instead on $this because you need to pass the $sequence intervals to the function.
Reduce seems that should work only on a single sequence, not 2.
Maybe it would be better to create another function or to add optional arguments to reduce() that can be passed used with the custom function.

**Comment by nyamsprod** [comment]
> What I feel is wrong is the use of reduce() called on the $sequence instead on $this 

The reduce method apply a function to each item of a collection. 

What Sequence::substract does from the examples and from your own implementation is essentially substracting EACH intervals from the provided sequence.
After each substract you may endup with a brand new sequence. The only data which does not change during the whole iteration is the provided sequence to substract AND NOT the soure Sequence. So to me it's normal that we use the reduce method of the only invariant data AKA the substract sequence 😉 

TL;DR:  the correct reduce is use here.

**Comment by Bibendus83** [comment]
That's the point, the logic of subtracting SequenceB from sequenceA means that substract should be called on A with argument B.
However the logic inverts when you call reduce on B passing A as argument and it feels weird to me. 

For readability I would have preferred something like this:
`return $this->reduce([$sequence, 'substractOne'], $this);`
but I have no idea if it would work

**Comment by nyamsprod** [comment]
the logic is not inverted ... what is reduce is sequenceA according to sequenceB but as soon as sequenceA is reduced even for one element it ceased to be sequenceA it becomes something else hence the only parameter which stay unchanged during the whole operation is indeed sequenceB ... meaning that you can only iterate safely on sequenceB but not on sequenceA ... so the choice is indeed logic not "natural" if by natural we are talking about naturally thinking we should iterate over sequenceA ... which is wrong

**Comment by Bibendus83** [comment]
It's just a readability issue for me, I know it would not work inverting the variables.

What about the possibility of passing custom additional arguments to reduce()?
In this case it would become
`return $this->reduce([$this, 'substractOne'], $this, $sequence);`

**Comment by nyamsprod** [comment]
what would be the point of passing these value around if they are not used ❓
A callable can take optional value like this 

```php
function ($foo, $bar) use ($baz) {
}
```

So you could but it does not mean that you should 😄 in this case. Also the last foreach could be rewritten to use reduce on the sequence but it this case it would **reduce** the readability IMHO

**Comment by Bibendus83** [comment]
Why do you say that these values are not used, in this case $sequence would be used and would make more sense as an argument instead of as the main object.

**Comment by nyamsprod** [comment]
I need to see your alternate implementation before making a judgement about it. If it is simpler than mine I'll gladly changed my POV 😉 .

**Comment by Bibendus83** [comment]
To use "use" in the callable as you said should I need to declare the function inside the substract method?

**Comment by nyamsprod** [comment]
I don't know where you want to declare your function but it should be declare before usage and once the data you want to associate through use is already defined otherwise it will failed

**Comment by nyamsprod** [comment]
for instance the substractOne method could be rewritten as follow:
```php
    /**
     * Substract an Interval from a Sequence.
     */
    private function substractOne(Sequence $sequence, Period $interval): self
    {
        if ($sequence->isEmpty()) {
            return $sequence;
        }

        $reducer = function (Sequence $sequence, Period $period) use ($interval) {
            $substract = $period->substract($interval);
            if (!$substract->isEmpty()) {
                $sequence->push(...$substract);
            }

            return $sequence;
        };
        
        return $sequence->reduce($reducer, new self());
    }
```

**Comment by nyamsprod** [comment]
@Bibendus83 any news on you alternate implementation 🤔 If not I may release the current implementation.

**Comment by Bibendus83** [comment]
Go for it, I don't have the time to work on it right now.
I'll see in the future what I can do.
